kubernetes/samples/scripts/domain-lifecycle/startServer.sh

```diff
@@ -0,0 +1,255 @@
+validateServerAndFindCluster "${domainUid}" "${domainNamespace}" "${serverName}" isValidServer clusterName
+if [ "${isValidServer}" != 'true' ]; then
+  printError "Server ${serverName} is not part of any cluster and it's not an independent managed server. Please make sure that server name specified is correct."
+  exit 1
+fi
+
+getClusterPolicy "${domainJson}" "${clusterName}" clusterPolicy
+if [ "${clusterPolicy}" == 'NEVER' ]; then
+  printError "Cannot start server '${serverName}', the server's parent cluster '.spec.clusters[?(clusterName=\"${clusterName}\"].serverStartPolicy' in the domain resource is set to 'NEVER'."
+  exit 1
+fi
+
+getDomainPolicy "${domainJson}" domainPolicy
+if [ "${domainPolicy}" == 'NEVER' ]; then
+  printError "Cannot start server '${serverName}', the .spec.serverStartPolicy in the domain resource is set to 'NEVER'."
+  exit 1
+fi
+
+getEffectivePolicy "${domainJson}" "${serverName}" "${clusterName}" effectivePolicy
+if [ -n "${clusterName}" ]; then
+  # Server is part of a cluster, check currently started servers
+  checkStartedServers "${domainJson}" "${serverName}" "${clusterName}" "${withReplicas}" "${withPolicy}" serverStarted
+  if [[ ${effectivePolicy} == "IF_NEEDED" && ${serverStarted} == "true" ]]; then
+    printInfo "No changes needed, exiting. The server should be already started or it's in the process of starting. The start policy for server ${serverName} is ${effectivePolicy} and server is chosen to be started based on current replica count."
+    exit 0
+  elif [[ "${effectivePolicy}" == "ALWAYS" && ${serverStarted} == "true" ]]; then
+    printInfo "No changes needed, exiting. The server should be already started or it's in the process of starting. The start policy for server ${serverName} is ${effectivePolicy}."
+    exit 0
+  fi
+else 
+  # Server is an independent managed server. 
+  if [[ "${effectivePolicy}" == "ALWAYS" || "${effectivePolicy}" == "IF_NEEDED" ]]; then
+    printInfo "No changes needed, exiting. The server should be already started or it's in the process of starting. The start policy for server ${serverName} is ${effectivePolicy}."
+    exit 0
+  fi
+fi
+
+getServerPolicy "${domainJson}" "${serverName}" managedServerPolicy
+targetPolicy="ALWAYS"
+createServerStartPolicyPatch "${domainJson}" "${serverName}" "${targetPolicy}" alwaysStartPolicyPatch 
+
+# if server is part of a cluster and replica count will increase
+if [[ -n ${clusterName} && "${keepReplicaConstant}" != 'true' ]]; then
+  #check if server starts by increasing replicas and unsetting policy
+  withReplicas="INCREASED"
+  withPolicy="UNSET"
+  checkStartedServers "${domainJson}" "${serverName}" "${clusterName}" "${withReplicas}" "${withPolicy}" startsByReplicaIncreaseAndPolicyUnset
+  operation="INCREMENT"
+  createReplicaPatch "${domainJson}" "${clusterName}" "${operation}" incrementReplicaPatch replicaCount
+  if [ "${incrementReplicaPatch}" == "MAX_REPLICA_COUNT_EXCEEDED" ]; then 
```

Comment:
If the replica count is already maxed out, why not just continue and start the server by unsetting it (e.g. implicitly bringing it to IF_NEEDED)?

Reply:
This code path for the default case where script increases the replica count. I'm not sure if unsetting policy (bringing it to IF_NEEDED) will start the server. The user can use '-k' to keep replica constant and start the server by unsetting policy. For e.g. if server1 to server4 are running and max replica count is 4. If user wants to start server5 and policy is NEVER, unsetting policy will not start server. Similarly if server5 policy is already IF_NEEDED then it can't be started without increasing replicas.

Reply:
As discussed, the assumption is that this case will be caught by previous validation check. I have removed the check.